synapse/storage/monthly_active_users.py

```diff
@@ -167,12 +183,18 @@ def get_registered_reserved_users_count(self):
             Defered[int]: Number of real reserved users
         """
         count = 0
+        users = ()
         for tp in self.hs.config.mau_limits_reserved_threepids:
             user_id = yield self.hs.get_datastore().get_user_id_by_threepid(
                 tp["medium"], tp["address"]
             )
             if user_id:
                 count = count + 1
+                users = users + (user_id,)
+
+        # Update reserved_users to ensure it stays in sync, this is important
+        # for reaping.
+        self.reserved_users = users
```

Comment:
This feels wrong, as there's nothing to say that `get_registered_reserved_users_count` is going to get called to make this accurate. Can't we query the database to get the `reserved_users` when we need it? (And potentially cache it?)

Reply:
Done, opted not to cache because method is called every hour and ought to be light.